### PR TITLE
Firewall UI

### PR DESCRIPTION
**Pull request by mridulnagpal**
#489

### CODE REVIEW

File: plinth/modules/firewall/templates/firewall.html [hunk 1]
```diff
@@ -40,48 +40,54 @@
...
-                        {% trans "Enabled" %}</span>
-                    {% else %}
-                      <span class='label label-warning'>
-                        {% trans "Disabled" %}</span>
-                    {% endif %}
-                  </td>
-                </tr>
-
-                {% for port in service.ports %}
+              {% if service.is_enabled %}
```
Comment (jvalleroy) [inline plinth/modules/firewall/templates/firewall.html]: My original idea was not to hide services that are disabled, but services that had never been setup.

File: plinth/modules/firewall/templates/firewall.html [hunk 2]
```diff
@@ -40,48 +40,54 @@
...
+                          {% trans "Enabled" %}</span>
                       {% else %}
-                        <span class='label label-danger'>
-                          {% trans "Blocked" %}</span>
+                        <span class='label label-warning'>
+                          {% trans "Disabled" %}</span>
                       {% endif %}
                     </td>
+                    <td>
+                      <button data-toggle="collapse" data-target=".{{service.service_id}}" class="btn btn-primary pull-right">View Ports</button>
```
Comment (jvalleroy) [inline plinth/modules/firewall/templates/firewall.html]: I wonder if we could streamline this a little. Could the "View Ports" button be made into a [+] icon to the left of the service name?

**Comment by mridulnagpal** [comment]
@jvalleroy How do I check if the app was never set up ??

**Comment by jvalleroy** [comment]
@mridulnagpal You can use setup.Helper.get_setup_version():
https://github.com/freedombox/Plinth/blob/master/plinth/setup.py#L141

Each module has its own instance of the setup.Helper class, called module.setup_helper:
https://github.com/freedombox/Plinth/blob/master/plinth/setup.py#L163

**Comment by mridulnagpal** [comment]
@jvalleroy I tried alot but couldn't find out what to do could you give me some more hints??

**Comment by jvalleroy** [comment]
Maybe this approach will work...

Currently, each module registers its service, when it is initialized. For example:
https://github.com/freedombox/Plinth/blob/master/plinth/modules/quassel/__init__.py#L69
This happens even if the module has not been setup yet.

Later, when the module setup is complete, it just notifies that its service is enabled:
https://github.com/freedombox/Plinth/blob/master/plinth/modules/quassel/__init__.py#L86

What I'm thinking is, maybe we don't always register the service in init(). Only register the service if the module has been setup. Then, setup() should register the service if it's not already registered (check if global service is None).

To check if the current module has been setup, this should work within a module's init():
```
    setup_helper = globals()['setup_helper']
    if setup_helper.get_state() != 'needs-setup':
        ...
```

**Comment by mridulnagpal** [comment]
Then how do I check if it is registered in the firewall template?

**Comment by jvalleroy** [comment]
@mridulnagpal With the approach I'm suggesting, you wouldn't need to check that in the template. Because the firewall view gets the list of services from the service module:
https://github.com/freedombox/Plinth/blob/master/plinth/modules/firewall/views.py#L43
So the services that don't get registered, will just not be included in this list.

**Comment by mridulnagpal** [comment]
@jvalleroy Please review the changes

**Comment by jvalleroy** [comment]
I've merged this, and added some code to the setup() to ensure that the service gets registered.

**Pull request closed without merging**